ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/RetryingGetBlockFromPeersTask.java

```diff
@@ -0,0 +1,98 @@
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.eth.manager.EthContext;
+import org.hyperledger.besu.ethereum.eth.manager.EthPeer;
+import org.hyperledger.besu.ethereum.eth.manager.exceptions.IncompleteResultsException;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class RetryingGetBlockFromPeersTask
+    extends AbstractRetryingPeerTask<AbstractPeerTask.PeerTaskResult<Block>> {
+
+  private static final int DEFAULT_MAX_RETRIES = 5;
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final ProtocolSchedule protocolSchedule;
+  private final List<EthPeer> peers;
+  private final Optional<Hash> blockHash;
+  private final long blockNumber;
+
+  public RetryingGetBlockFromPeersTask(
+      final EthContext ethContext,
+      final ProtocolSchedule protocolSchedule,
+      final List<EthPeer> peers,
+      final MetricsSystem metricsSystem,
+      final Optional<Hash> blockHash,
+      final long blockNumber,
+      final int maxRetries) {
+    super(ethContext, maxRetries, Objects::isNull, metricsSystem);
+    this.protocolSchedule = protocolSchedule;
+    this.peers = peers;
+    this.blockHash = blockHash;
+    this.blockNumber = blockNumber;
+  }
+
+  public static RetryingGetBlockFromPeersTask create(
+      final EthContext ethContext,
+      final ProtocolSchedule protocolSchedule,
+      final List<EthPeer> peers,
+      final Optional<Hash> hash,
+      final long blockNumber,
+      final MetricsSystem metricsSystem) {
+    return new RetryingGetBlockFromPeersTask(
+        ethContext, protocolSchedule, peers, metricsSystem, hash, blockNumber, DEFAULT_MAX_RETRIES);
+  }
+
+  @Override
+  protected CompletableFuture<AbstractPeerTask.PeerTaskResult<Block>> executePeerTask(
+      final Optional<EthPeer> assignedPeer) {
+    final GetBlockFromPeersTask getHeadersTask =
+        GetBlockFromPeersTask.create(
+            peers, protocolSchedule, getEthContext(), blockHash, blockNumber, getMetricsSystem());
+    return executeSubTask(getHeadersTask::run)
+        .thenApply(
+            peerResult -> {
+              result.complete(peerResult);
+              return peerResult;
+            });
+  }
+
+  @Override
+  protected boolean isRetryableError(final Throwable error) {
+    return super.isRetryableError(error) || error instanceof IncompleteResultsException;
+  }
+
+  @Override
+  protected void handleTaskError(final Throwable error) {
+    if (getRetryCount() < getMaxRetries()) {
+      LOG.info(
+          "Failed to get block with hash {} and number {} retrying later", blockHash, blockNumber);
+    }
+    super.handleTaskError(error);
+  }
```

Comment:
What is our recovery path if we hit max retries?  is there any mechanism to periodically retry fetching blocks or are we just stuck?

Reply:
I voluntarily left a maximum because we may actually have a bad block that has been broadcast and that we will never find. This could be a way to back us up if a peers broadcast us a huge list of n+1  blocks IMO

I think rather than trying stubbornly to find a block. the best way is to change peers at some point if we are no longer moving forward. This could be discussed as a follow-up to this PR in a second time